Use monomorphic let constraints instead of def

jstolarek opened the pull request:
This PR implements let_mono constraints in place of `def` and monomorphic predicates. One test is currently failing - @frank-emrich, can you take a look and tell me your thoughts? I had to comment out an assertion in `Unifier`, line 108. Apparently in some situations we are attempting to skolemize monomorphic variables. I haven't investigated yet but we need to take a look at this since it might indicate a bug. Overall constraint generation code in the client got significantly shorter, which is a huge improvement. There's no extra complexity in the solver itself, though obviously some changes were required.

jstolarek committed:
Parametrize CLet by its type (generalising or monomorphising)

jstolarek committed:
Expose let type in high-level interface

jstolarek committed:
Implement C-Abs based on let_mono

jstolarek committed:
Implement C-AbsAscribe based on let_mono

jstolarek committed:
Remove exists_sigm use optional argument instead

jstolarek committed:
Merge constraint generation for annotated and unannotated lambdas

jstolarek committed:
WIP implementation of let restriction using let_mono

jstolarek committed:
Merge constraint generation for annotated and unannotated lets

jstolarek committed:
Further merge constraint generation for lets

jstolarek committed:
Add functions for setting monomorphic restriction

jstolarek committed:
Remove unused CDef constraint and PMono* predicates

jstolarek committed:
Code shuffle

jstolarek committed:
Remove redundant is_gval argument

jstolarek committed:
Propagate monomorphic restriction when instantiating quantifiers

jstolarek committed:
Rename let1 to let1_gen

jstolarek committed:
Add debugging info

jstolarek commented:
Need to record some thoughts since not everything seems to be right. Consider the program:
```
let f = id id in f ~id
```
Currently `f` receives a type scheme `forall a. a -> a` and we enforce correct behaviour by monomorphising quantifier `a`, which prevents it from being instantiated with a polymorphic type. This works as intended but is not correct since `f` should not be generalised at all. It should have the type `a -> a` where free variable `a` is restricted to be monomorphic. However, any of my attempts to fix it messes up something else. My understanding is that if we have a let_mono construct of the form `let_mono x = M in N` then:

  1. expression `M` is not generalised, which means that free type variables in the inferred type of M remain unbound.  These variables however must be made monomorphic.
  2. expression `M` is allowed to have a polymorphic type, in which case any top level quantifiers should not be monomorphised.
  3. there should be no generalisable variables to be inserted in a resulting System F term.
  4. It is important to also monomorphize free type variables. This ensures correct implementation of monomorphic restriction on unannotated lambda binders in examples such as `\x. choose ~id x`, where `x` inferred type is just a type variable.

I believe there is also a connection between 1 and 3: any determined generalizable variables should be removed from the quantifiers of an inferred type scheme.

However, implementing the above leads to problems with unbound unification variables, as originally speculated in #24. These are manifested in the failure of `nested_forall_inst_3`, which produces:

```
Λ61. let id = Λ38. λ(x : 38). x in
let auto' = Λ42. λ(x : ∀ [25]. 25 → 25). x [42 → 42] (x [42]) in
let f = Λ1. id [(∀ [3]. 3 → 3) → 1 → 1] in
let g = f [55] (id [(∀ [58]. 58 → 58) → 55 → 55] (auto' [55])) in
g [61] id
```

with variable `55` being unbound. I believe a correct program should look like this:

```
Λ61. let id = Λ38. λ(x : 38). x in
let auto' = Λ42. λ(x : ∀ [25]. 25 → 25). x [42 → 42] (x [42]) in
let f = Λ1. id [(∀ [3]. 3 → 3) → 1 → 1] in
let g = f [61] (id [(∀ [58]. 58 → 58) → 61 → 61] (auto' [61])) in
g id
```

For now I pushed this on a spearate branch `let_mono_nogen`.

jstolarek committed:
Revert "More accurate pretty-printing"

This reverts commit c54f148b7d49c67d8b078cc53b3925f16a869e8c.

jstolarek committed:
Revert "More accurate pretty-printing"

This reverts commit b0578af595ca7ad19c1fcb948b08eb53b0a0b21c.

jstolarek committed:
Print variable restrictions when printing a pool

jstolarek committed:
Remove redundant monomorphisation

jstolarek committed:
Minor tweaks, extra assertions

frank-emrich commented:
I haven't looked at the code yet, but here are some questions:

> Currently f receives a type scheme forall a. a -> a and we enforce correct behaviour by monomorphising quantifier a, which prevents it from being instantiated with a polymorphic type. This works as intended but is not correct since f should not be generalised at all.

One way of looking at it, though, is that we can generalize f and then immediately instantiate those quantifiers we just generalized with fresh monomorphic unification variables.


> 2. expression M is allowed to have a polymorphic type, in which case any top level quantifiers should not be monomorphised.

This makes it seem like we currently track for quantifiers whether they are monomorpic or polymorphic. This shouldn't be necessary, as quantified type variables should always be considered polymorphic. In other words, *use sites* determine whether quantified variables should be instantiated monomorphically or polymorphically.

> However, implementing the above leads to problems with unbound unification variables, as originally speculated in #24. These are manifested in the failure of nested_forall_inst_3, which produces:

I dont understand the expected result type ∀ a. a → a for `nested_forall_inst_3`. Why is it polymorphic? I vaguely remember that the implemenation generalizes everything at the toplevel. But does this generalization obey the value restriction or not? If it does, then the expected type should just be a -> a for some monomorphic, but otherwise unconstrained type variable a.

But in any case, there is definitely something wrong with the generated System F term, it should be using type variable 61 throught, as you've done in your corrected version (I assume the few occurences of 55 that are left in your corrected version were also meant to be 61s). This should be independent from whether or not there is a  Λ61 at the outermost level or not, so it's not really affected by my earlier question about toplevel generalization

jstolarek committed:
Add tests for quantifier ordering

jstolarek committed:
Change test specification

jstolarek committed:
Obey value restriction at the top-level let

jstolarek committed:
Print representativity information

jstolarek committed:
Print representativity information with Debug.hard

jstolarek committed:
Don't base generalization on representativity

Closes #34

jstolarek committed:
Test comments

jstolarek commented:
> This makes it seem like we currently track for quantifiers whether they are monomorpic or polymorphic.

This is exactly how things work at the moment. For example if we have `let f = id id in f ~id` we infer the type of `f` to be `forall a. a -> a` and mark quantifier `a` as being monomorphic. I realize this isn't exactly how we specified things in the PLDI paper but I believe the end result is the same.

> I dont understand the expected result type ∀ a. a → a for nested_forall_inst_3. Why is it polymorphic? I vaguely remember that the implemenation generalizes everything at the toplevel.

Yes, correct. In the test specification I write `[∀ a.] a → a`, where square brackets indicate that the quantifier comes from top-level generalisation. I forgot to put the brackets it in this case. This is now fixed.

> But does this generalization obey the value restriction or not? 

I changed it today to obey the value restriction, but at the moment this doesn't lead to any noticeable differences due to the way we handle monomorphic lets, described above.

I will merge this PR now since it achieves the main goal of eliminating def constraints and correctly implementing the value restriction using monomorphic lets. I will leave the task of changing the internal behaviour of monomorphic lets for the future.
